api/tests/conftest.py

```diff
@@ -0,0 +1,96 @@
+# Update path so the code from the src directory gets imported correctly for
+# tests without needing to modify module imports in src
+import sys
+sys.path.append('/meep/api/src')
+
+import pytest
+from app import create_app
+from models import User, Role, Project, db, ProjectType
+
+
+@pytest.fixture(scope='module')
```

Comment:
I think we want ```scope='function'``` in the fixtures that instantiate models. This will prevent changes made to a model in one test from affecting the input data to a different test in the same module, and will ensure that the tests are independent of the order in which they are run. As an example, I added a fixture to conftest.py:

```python
@pytest.fixture(scope='module')
def breakfast():
    return ['eggs', 'waffles']
```

and then two tests to test_models.py:

```python
def test_add_bacon(breakfast):
    assert breakfast == ['eggs', 'waffles']
    breakfast.append('bacon')
    assert breakfast == ['eggs', 'waffles', 'bacon']


def test_add_pancakes(breakfast):
    assert breakfast == ['eggs', 'waffles']
    breakfast.append('pancakes')
    assert breakfast == ['eggs', 'waffles', 'pancakes']
```

the output after running the tests was

```
======================================= test session starts ========================================
platform linux -- Python 3.7.3, pytest-4.6.2, py-1.8.0, pluggy-0.12.0
rootdir: /meep/api/src, inifile: pytest.ini, testpaths: /meep/api/tests
collected 7 items

functional/test_api.py ..                                                                    [ 28%]
unit/test_models.py ....F                                                                    [100%]

============================================= FAILURES =============================================
________________________________________ test_add_pancakes _________________________________________

breakfast = ['eggs', 'waffles', 'bacon']

    def test_add_pancakes(breakfast):
>       assert breakfast == ['eggs', 'waffles']
E       AssertionError: assert ['eggs', 'waffles', 'bacon'] == ['eggs', 'waffles']
E         Left contains one more item: 'bacon'
E         Use -v to get the full diff

../tests/unit/test_models.py:41: AssertionError
```

so the changes to the fixture from test_add_bacon are propagating to test_add_pancakes.

Reply:
Thanks for that explanation. Still learning pytest but that makes perfect sense. Updated those changes.